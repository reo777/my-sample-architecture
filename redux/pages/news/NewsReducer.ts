import { combineReducers, createReducer } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';
import type { TypedUseSelectorHook } from 'react-redux';

import { getNewsItems } from './newsActions';

interface NewsItemState {
  id: number;
  title: string;
  photoUrls: string[];
  publishedAt: string;
  isNew: boolean;
  isLiked: boolean;
}

const initialState: NewsItemState[] = [
  {
    id: 0,
    title: '',
    photoUrls: [''],
    publishedAt: '',
    isNew: true,
    isLiked: true,
  },
];

const newsReducer = createReducer(initialState, builder =>
  builder.addCase(getNewsItems, (draft, action) => {
    draft;
  }),
);

export const newsRootReducer = {
  newsItems: newsReducer,
};

const localReducer = combineReducers(newsRootReducer);
type NewsRootState = ReturnType<typeof localReducer>;
export const useNewsSelector: TypedUseSelectorHook<NewsRootState> = useSelector;
