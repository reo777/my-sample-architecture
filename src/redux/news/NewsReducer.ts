import { combineReducers, createReducer } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';
import type { TypedUseSelectorHook } from 'react-redux';

import {
  updateNewsItemsAction,
  updateLikeAction,
  deleteLikeAction,
} from './NewsActions';

export interface NewsItemState {
  id: number;
  title: string;
  photoUrls: string[];
  publishedAt: string;
  isNew: boolean;
  isLiked: boolean;
}

const initialState: { newsItems: NewsItemState[] } = { newsItems: [] };

export const updatedLikeItems = (
  newsItems: NewsItemState[],
  action: {
    payload: {
      id: number;
    };
  },
) => {
  return newsItems.map(newsItem => {
    if (newsItem.id === action.payload.id) {
      return {
        ...newsItem,
        isLiked: true,
      };
    }
    return newsItem;
  });
};

export const deletedLikeItems = (
  newsItems: NewsItemState[],
  action: {
    payload: {
      id: number;
    };
  },
) => {
  return newsItems.map(newsItem => {
    if (newsItem.id === action.payload.id) {
      return {
        ...newsItem,
        isLiked: false,
      };
    }
    return newsItem;
  });
};

const newsReducer = createReducer(initialState, builder => {
  builder
    .addCase(updateNewsItemsAction, (draft, action) => {
      draft.newsItems = action.payload.newsItems;
    })

    .addCase(updateLikeAction, (draft, action) => {
      draft.newsItems = updatedLikeItems(draft.newsItems, action);
    })

    .addCase(deleteLikeAction, (draft, action) => {
      draft.newsItems = deletedLikeItems(draft.newsItems, action);
    });
});

export const newsRootReducer = {
  news: newsReducer,
};

const localReducer = combineReducers(newsRootReducer);

type NewsRootState = ReturnType<typeof localReducer>;
export const useNewsSelector: TypedUseSelectorHook<NewsRootState> = useSelector;
