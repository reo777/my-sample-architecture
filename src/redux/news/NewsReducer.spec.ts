import { updatedLikeItems, deletedLikeItems } from './NewsReducer';
describe('NewsReducer', () => {
  const draftUpdateLike = [
    {
      id: 0,
      title: 'test news item',
      photoUrls: [''],
      publishedAt: '2021年5月8日',
      isNew: true,
      isLiked: false,
    },
  ];
  const draftDeleteLike = [
    {
      id: 0,
      title: 'test news item',
      photoUrls: [''],
      publishedAt: '2021年5月8日',
      isNew: true,
      isLiked: false,
    },
  ];

  it('updatedLikeItems', () => {
    const action = { payload: { id: 0 } };

    const result = updatedLikeItems(draftUpdateLike, action);

    expect(result).toMatchSnapshot();
  });

  it('deletedLikeItems', () => {
    const action = { payload: { id: 0 } };

    const result = deletedLikeItems(draftDeleteLike, action);

    expect(result).toMatchSnapshot();
  });
});
