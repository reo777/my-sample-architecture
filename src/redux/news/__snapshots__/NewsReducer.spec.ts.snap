// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsReducer deletedLikeItems 1`] = `
Array [
  Object {
    "id": 0,
    "isLiked": false,
    "isNew": true,
    "photoUrls": Array [
      "",
    ],
    "publishedAt": "2021年5月8日",
    "title": "test news item",
  },
]
`;

exports[`NewsReducer updatedLikeItems 1`] = `
Array [
  Object {
    "id": 0,
    "isLiked": true,
    "isNew": true,
    "photoUrls": Array [
      "",
    ],
    "publishedAt": "2021年5月8日",
    "title": "test news item",
  },
]
`;
